clients/drcachesim/simulator/caching_device.cpp

```diff
@@ -227,18 +267,101 @@ caching_device_t::invalidate(const addr_t tag)
         if (cache_block.tag == tag) {
             cache_block.tag = TAG_INVALID;
             cache_block.counter = 0;
-            stats->invalidate();
+            stats->invalidate(invalidation_type_);
             // Invalidate last_tag if it was this tag.
             if (last_tag == tag) {
                 last_tag = TAG_INVALID;
             }
             // Invalidate the block in the children's caches.
-            if (inclusive && !children.empty()) {
+            if (invalidation_type_ == INVALIDATION_INCLUSIVE && inclusive &&
+                !children.empty()) {
                 for (auto &child : children) {
-                    child->invalidate(tag);
+                    child->invalidate(tag, invalidation_type_);
                 }
             }
             break;
         }
     }
+    // If this is a coherence invalidation, we must invalidate children caches.
+    if (invalidation_type_ == INVALIDATION_COHERENCE && !children.empty()) {
+        for (auto &child : children) {
+            child->invalidate(tag, invalidation_type_);
+        }
+    }
+}
+
+// This function checks if this cache or any child caches hold a tag.
+bool
+caching_device_t::check_tags(const addr_t &tag_in)
+{
+    int block_idx = compute_block_idx(tag_in);
+    for (int way = 0; way < associativity; way++) {
+        if (get_caching_device_block(block_idx, way).tag == tag_in) {
+            return true;
+        }
+    }
+    if (children.empty()) {
+        return false;
+    } else {
```

Comment:
nit: no need for this else -- you can reduce nesting.

Reply:
Fixed.